JavaScript/Angular.md

```diff
@@ -0,0 +1,1551 @@
+    - some-other.component.ts
+      > `export class SomeOtherComponent implements ISomeOtherComponent`
+    - some-other.component.spec.ts
+    - some-other.component.interface.ts
+      > `export interface ISomeOtherComponent`
+
+      > interface should include all `@Input()`s and `@Output()`s
+    - some-other.testing.component.ts
+      > `export class SomeOtherTestingComponent implements ISomeOtherComponent`
+
+      > **important** - should have the same selector as `SomeOtherComponent`
+    - some-other.testing.module.ts
+      > `export class SomeOtherTestingModule`
+
+      > should declare and export `SomeOtherTestingComponent`
+
+Make sure to exclude `*.testing.*` files from code coverage reports in Jasmine (`angular.json` `codeCoverageExclude`) and any other reporting tools you might be using (like SonarQube, for example).
+
+Test should look like this:
+
+```ts
+let component: ComponentToBeTested;
+let fixture: ComponentFixture<ComponentToBeTested>;
+
+beforeEach(async(() => {
+  TestBed.configureTestingModule({
+    imports: [SomeOtherTestingModule], // use testing child component
+    declarations: [ComponentToBeTested],
+    providers: [
+      { provide: UserService, useClass: UserTestingService }, // use testing service
+    ],
+  })
+  .compileComponents();
+}));
+
+beforeEach(() => {
+  fixture = TestBed.createComponent(ComponentToBeTested);
+  component = fixture.componentInstance;
+  fixture.detectChanges();
+});
+
+it('should create', () => {
+  expect(component).toBeTruthy();
+});
+```
+
+### Testing component inputs
+
+If you want to test a component under different conditions, you will probably have to change some of its `@Input()` values and check if it emits `@Output()`s when expected.
+
+To change components input, you can simply:
+
+```ts
+@Component(...)
+class SomeComponent {
+  @Input() someInput: string;
+}
+```
+
+```ts
+component.someInput = 'foo';
+fixture.detectChanges(); // To update the view
+```
+
+If you are doing stuff in `ngOnChanges`, you will have to call it manually since in tests `ngOnChanges` is not called automatically during programmatic input changes.
+
+```ts
+component.someInput = 'foo';
+component.ngOnChanges({ someInput: { currentValue: 'foo' } } as any);
+fixture.detectChanges();
+```
+
+### Testing component outputs
+
+To test component outputs, simply spy on output emit function:
+
+```ts
+@Component(
+  template: `<button (click)="onButtonClicked()">`
+)
+export class ComponentWithOutput {
+  @Output() public buttonClicked: EventEmitter<void> = new EventEmitter();
+
+  public onButtonClicked(): void {
+    this.buttonClicked.emit();
+  }
+}
+```
+
+```ts
+spyOn(component.buttonClicked, 'emit');
+
+expect(component.buttonClicked.emit).not.toHaveBeenCalled();
+
+fixture.debugElement.query(By.css('button')).nativeElement.click();
+
+expect(component.buttonClicked.emit).toHaveBeenCalled();
+```
+
+### Testing components with `OnPush` change detection
```

Comment:
Maybe could be useful to add simple example for testing components containing `ng-content` with wrapper?

Reply:
Added